ecosystem/sf-indexer/aptos-substreams/src/token_converter.rs

```diff
@@ -0,0 +1,369 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{Context, Result};
+use aptos_protos::{
+    block_output::v1::{MoveResourceOutput, TableItemOutput},
+    tokens::v1::{CollectionData, Token, TokenData, TokenDataId, TokenId},
+};
+use std::collections::HashMap;
+
+/// Mapping from table handle to owner type, including type of the table (AKA resource type)
+pub fn get_table_handle_to_owner(
+    resource: &MoveResourceOutput,
+    txn_version: u64,
+) -> anyhow::Result<Option<HashMap<String, (String, String)>>> {
+    match resource.type_str.as_str() {
+        "0x3::token::Collections" => {
+            let address = &resource.address;
+            let data: serde_json::Value = serde_json::from_str(&resource.data)?;
+            let collection_handle = data["data"]["collection_data"]["handle"]
+                .as_str()
+                .map(|s| s.to_string())
+                .context(format!(
+                    "version {} failed! collection data handle must be present {:?}",
+                    txn_version, data
+                ))?;
+
+            Ok(Some(HashMap::from([(
+                standardize_handle(&collection_handle),
+                (address.clone(), resource.type_str.clone()),
+            )])))
+        }
+        "0x3::token::TokenStore" => {
+            let address = &resource.address;
+            let data: serde_json::Value = serde_json::from_str(&resource.data)?;
+            let token_store_handle = data["data"]["tokens"]["handle"]
+                .as_str()
+                .map(|s| s.to_string())
+                .context(format!(
+                    "version {} failed! token store data handle must be present {:?}",
+                    txn_version, data
+                ))?;
+            Ok(Some(HashMap::from([(
+                standardize_handle(&token_store_handle),
+                (address.clone(), resource.type_str.clone()),
+            )])))
+        }
+        _ => Ok(None),
+    }
+}
+
+fn standardize_handle(handle: &String) -> String {
+    format!("0x{}", &handle[2..].trim_start_matches('0'))
+}
+
+pub fn get_token(
+    table_item: &TableItemOutput,
+    txn_version: u64,
+    table_handle_to_owner: &HashMap<String, (String, String)>,
+) -> anyhow::Result<Option<Token>> {
+    if table_item.key_type != "0x3::token::TokenId" {
+        return Ok(None);
+    }
+    let table_handle = standardize_handle(&table_item.handle);
+    let (owner_address, table_type) = table_handle_to_owner
```

Comment:
Is this the table value type?

Reply:
No it's the resource type and if the resource is a table i'm calling it table type.